src/main/groovy/pl/allegro/tech/build/axion/release/domain/Releaser.groovy

```diff
@@ -35,12 +37,33 @@ class Releaser {
             repository.tag(tagName)
 
             hooksRunner.runPostReleaseHooks(rules.hooks, rules, versionContext, version)
+            return Optional.of(tagName)
         } else {
             logger.quiet("Working on released version ${version}, nothing to release.")
+            return Optional.empty()
+        }
+    }
+
+    void releaseAndPush(Properties rules) {
+        def releasedTagName = release(rules)
+
+        try {
+            pushRelease()
+        } catch (ScmPushFailedException | GitAPIException e) {
+            logger.quiet("Exception occurred during pushing the release: ${e.message}.")
+            releasedTagName.ifPresent{ tagName ->
+                logger.quiet("Removing tag: $tagName.")
+                rollbackRelease(tagName)
+            }
+            throw e
         }
     }
 
     void pushRelease() {
         repository.push()
```

Comment:
exception handling logic is missing here - this method is called from `PushRelease` task and displayed output should be the same (only without rollback)

Reply:
I actually stopped myself from making too many changes to existing structures, because that would start an endless stream of "Oh and while I'm at it, I'll change this this and that too" and possibly ruin some design.

That being said, it sounds very reasonable to apply what you suggest here since it would be really awkward for those two methods to act differently in case of an error.